crates/lang_srv/src/server.rs

```diff
@@ -61,39 +67,124 @@ impl RocLs {
+            resolve_provider: Some(false),
+            trigger_characters: Some(vec![".".to_string()]),
+            //TODO: what is this?
+            all_commit_characters: None,
+            work_done_progress_options: WorkDoneProgressOptions {
+                work_done_progress: None,
+            },
+        };
         ServerCapabilities {
             text_document_sync: Some(text_document_sync),
             hover_provider: Some(hover_provider),
             definition_provider: Some(OneOf::Right(definition_provider)),
             document_formatting_provider: Some(OneOf::Right(document_formatting_provider)),
             semantic_tokens_provider: Some(semantic_tokens_provider),
+            completion_provider: Some(completion_provider),
             ..ServerCapabilities::default()
         }
     }
 
     /// Records a document content change.
     async fn change(&self, fi: Url, text: String, version: i32) {
-        self.registry()
-            .apply_change(DocumentChange::Modified(fi.clone(), text));
+        let updating_result = self.state.change(&fi, text, version).await;
 
-        let diagnostics = match std::panic::catch_unwind(|| self.registry().diagnostics(&fi)) {
-            Ok(ds) => ds,
-            Err(_) => return,
-        };
+        //The analysis task can be cancelled by another change coming in which will update the watched variable
+        if let Err(e) = updating_result {
+            debug!("cancelled change. Reason:{:?}", e);
+            return;
+        }
+        debug!("applied_change getting and returning diagnostics");
+
+        let diagnostics = self.state.registry.diagnostics(&fi).await;
 
         self.client
             .publish_diagnostics(fi, diagnostics, Some(version))
             .await;
     }
+}
+
+impl RocServerState {
+    pub fn new() -> RocServerState {
+        Self {
+            registry: Registry::default(),
+        }
+    }
+
+    async fn registry(&self) -> &Registry {
+        &self.registry
+    }
+
+    async fn close(&self, _fi: Url) {
+        ()
+    }
+
+    pub async fn change(
+        &self,
+        fi: &Url,
+        text: String,
+        version: i32,
+    ) -> std::result::Result<(), String> {
+        debug!("V{:?}:starting change", version);
+        let doc_info = DocInfo::new(fi.clone(), text, version);
+
+        self.registry
+            .apply_doc_info_changes(fi.clone(), doc_info.clone())
+            .await;
+
+        debug!(
+            "V{:?}:finished updating docinfo, starting analysis ",
+            version
+        );
+
+        let inner_ref = self;
+        let updating_result = async {
+            //This reduces wasted computation by waiting to allow a new change to come in and update the version before we check, but does delay the final analysis. Ideally this would be replaced with cancelling the analysis when a new one comes in.
+            tokio::time::sleep(Duration::from_millis(100)).await;
```

Comment:
I'm +1 to this, but can we do two things:

- I'd like to lower this to 35ms, if that sounds reasonable.
- Could we also make this parameterizable via an env var? For example, `ROCLS_DEBOUNCE_MS`. I think this would be useful in order to figure out how we can improve this over time.

Reply:
I'm happy to add the environment variable but I would prefer not to lower the debounce in fact I probably lean towards raising it to 150 milliseconds. 
I did some tests last night to see how fast I typed and on average during a word I hit around 10-15 keystrokes per second and 6 to 10 across multiple words.

35 milliseconds would simply be far too small a delay for the next change event to be able to intercept the analysis.

I don't think the difference between 100 milliseconds and 35 milliseconds is meaningful given the analysis itself takes close to a second

Reply:
ok, 100ms sounds good